File: fieldEncryption.js
const crypto = require('crypto');
const { wrapDataKey, unwrapDataKey } = require('../config/kms');

const ALGO = 'aes-256-gcm';

/**
 * Encrypts a plain object of string fields (e.g. { name, email }).
 * Generates one fresh Data Encryption Key (DEK) per call, encrypts each
 * field locally with it (fast, no network call per field), then sends
 * ONLY the 32-byte DEK to Cloud KMS to be wrapped. This is envelope
 * encryption — the pattern Google recommends for field-level encryption,
 * since it means one KMS API call per document instead of one per field.
 *
 * Returns { encrypted: { name: '<base64>', email: '<base64>' }, wrappedDek: '<base64>' }
 */
async function encryptFields(fields) {
  const dek = crypto.randomBytes(32);
  const encrypted = {};

  for (const [key, value] of Object.entries(fields)) {
    if (value === undefined || value === null || value === '') continue;

    const iv = crypto.randomBytes(12);
    const cipher = crypto.createCipheriv(ALGO, dek, iv);
    const ciphertext = Buffer.concat([cipher.update(String(value), 'utf8'), cipher.final()]);
    const authTag = cipher.getAuthTag();

    // Store iv + authTag + ciphertext together so decryption is self-contained
    encrypted[key] = Buffer.concat([iv, authTag, ciphertext]).toString('base64');
  }

  const wrappedDek = await wrapDataKey(dek);
  dek.fill(0); // wipe the plaintext DEK from memory once it's safely wrapped

  return { encrypted, wrappedDek: wrappedDek.toString('base64') };
}

/**
 * Reverses encryptFields(). Unwraps the DEK via KMS (1 call), then
 * decrypts every field locally with it.
 */
async function decryptFields(encrypted, wrappedDekBase64) {
  if (!wrappedDekBase64) return {};

  const wrappedDek = Buffer.from(wrappedDekBase64, 'base64');
  const dek = await unwrapDataKey(wrappedDek);

  const decrypted = {};
  for (const [key, value] of Object.entries(encrypted || {})) {
    if (!value) continue;

    const buf = Buffer.from(value, 'base64');
    const iv = buf.subarray(0, 12);
    const authTag = buf.subarray(12, 28);
    const ciphertext = buf.subarray(28);

    const decipher = crypto.createDecipheriv(ALGO, dek, iv);
    decipher.setAuthTag(authTag);
    const plaintext = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
    decrypted[key] = plaintext.toString('utf8');
  }

  return decrypted;
}

module.exports = { encryptFields, decryptFields };

File: field_encryption.py
"""Envelope encryption for individual fields (AES-256-GCM + Cloud KMS).

One fresh Data Encryption Key (DEK) is generated per call. Each field is
encrypted locally with that DEK (fast, no network call per field), then
only the 32-byte DEK is sent to Cloud KMS to be wrapped. This is the
pattern Google recommends for field-level encryption.

encrypted_fields structure stored in MongoDB:
{
    "name": "<base64(iv + authTag + ciphertext)>",
    "email": "<base64(iv + authTag + ciphertext)>",
    ...
}
wrapped_dek: "<base64>"  (stored alongside, one per document)
"""

import base64
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from kms import unwrap_data_key, wrap_data_key

_ALGO_NONCE_LEN = 12  # 96-bit nonce for AES-GCM


def _b64(buf: bytes) -> str:
    return base64.b64encode(buf).decode()


def _unb64(s: str) -> bytes:
    return base64.b64decode(s)


def encrypt_fields(fields: dict[str, str | None]) -> tuple[dict[str, str], str]:
    """Encrypt a dict of string fields.

    Returns (encrypted_dict, wrapped_dek_b64).
    """
    dek = os.urandom(32)
    aesgcm = AESGCM(dek)

    encrypted: dict[str, str] = {}
    for key, value in fields.items():
        if value is None or value == "":
            continue
        nonce = os.urandom(_ALGO_NONCE_LEN)
        ct = aesgcm.encrypt(nonce, value.encode("utf-8"), None)
        # Store nonce (12) + ciphertext (includes 16-byte GCM tag)
        encrypted[key] = _b64(nonce + ct)

    wrapped = wrap_data_key(dek)
    return encrypted, _b64(wrapped)


def decrypt_fields(encrypted: dict[str, str] | None, wrapped_dek_b64: str) -> dict[str, str]:
    """Decrypt a dict of fields previously encrypted with encrypt_fields."""
    if not wrapped_dek_b64 or not encrypted:
        return {}

    dek = unwrap_data_key(_unb64(wrapped_dek_b64))
    aesgcm = AESGCM(dek)

    decrypted: dict[str, str] = {}
    for key, value in encrypted.items():
        if not value:
            continue
        raw = _unb64(value)
        nonce = raw[:_ALGO_NONCE_LEN]
        ct = raw[_ALGO_NONCE_LEN:]
        pt = aesgcm.decrypt(nonce, ct, None)
        decrypted[key] = pt.decode("utf-8")

    return decrypted
